druid/src/widget/button.rs

```diff
@@ -13,146 +13,98 @@
+            // on HotChanged and active.
+            .on_click(|_, _, _| {})
     }
 
-    fn layout(
-        &mut self,
-        layout_ctx: &mut LayoutCtx,
-        bc: &BoxConstraints,
-        data: &T,
-        env: &Env,
-    ) -> Size {
-        bc.debug_check("Button");
-        let padding = Size::new(LABEL_INSETS.x_value(), LABEL_INSETS.y_value());
-        let label_bc = bc.shrink(padding).loosen();
-        self.label_size = self.label.layout(layout_ctx, &label_bc, data, env);
-        // HACK: to make sure we look okay at default sizes when beside a textbox,
-        // we make sure we will have at least the same height as the default textbox.
-        let min_height = env.get(theme::BORDERED_WIDGET_HEIGHT);
-
-        bc.constrain(Size::new(
-            self.label_size.width + padding.width,
-            (self.label_size.height + padding.height).max(min_height),
-        ))
+    /// Create a new button with a child widget for a label.
+    ///
+    /// The widget will receive padding and a styled background and border.If
+    /// you want a clickable widget without the styling, consider just using
+    /// `.on_click` without the Button.
+    pub fn with_child(child: impl Widget<T> + 'static) -> impl Widget<T> {
+        let painter = Self::painter();
+        child
+            .padding(LABEL_INSETS)
+            .background(painter)
+            // TODO: this is a hacky way to make sure the Painter is updated
+            // on HotChanged and active.
+            .on_click(|_, _, _| {})
     }
 
-    fn paint(&mut self, ctx: &mut PaintCtx, data: &T, env: &Env) {
-        let is_active = ctx.is_active();
-        let is_hot = ctx.is_hot();
-        let size = ctx.size();
-
-        let rounded_rect = Rect::from_origin_size(Point::ORIGIN, size)
-            .to_rounded_rect(env.get(theme::BUTTON_BORDER_RADIUS));
-
-        let bg_gradient = if is_active {
-            LinearGradient::new(
-                UnitPoint::TOP,
-                UnitPoint::BOTTOM,
-                (env.get(theme::BUTTON_LIGHT), env.get(theme::BUTTON_DARK)),
-            )
-        } else {
-            LinearGradient::new(
-                UnitPoint::TOP,
-                UnitPoint::BOTTOM,
-                (env.get(theme::BUTTON_DARK), env.get(theme::BUTTON_LIGHT)),
-            )
-        };
-
-        let border_color = if is_hot {
-            env.get(theme::BORDER_LIGHT)
-        } else {
-            env.get(theme::BORDER_DARK)
-        };
-
-        ctx.stroke(
-            rounded_rect,
-            &border_color,
-            env.get(theme::BUTTON_BORDER_WIDTH),
-        );
-
-        ctx.fill(rounded_rect, &bg_gradient);
-
-        let label_offset = (size.to_vec2() - self.label_size.to_vec2()) / 2.0;
-
-        ctx.with_save(|ctx| {
-            ctx.transform(Affine::translate(label_offset));
-            self.label.paint(ctx, data, env);
-        });
+    fn painter() -> Painter<T> {
```

Comment:
One thing I've been thinking about is that we could use `Painter` to do button styles? This is a bit tricky because we might want to be able to assume that a button has text, i.e. we might want a text button with different styles. Is there a way we can do this? Can we have a reusable 'button style', that works with the normal button API? I know this isn't really the scope for this PR, but I'm spitballing.

Reply:
It's not quite clear to me how, but I do think somehow Painter will give us a good way for styling Button.